lib/puppetserver/ca/sign_action.rb

```diff
@@ -0,0 +1,213 @@
+      def self.parser(parsed = {})
+        OptionParser.new do |opts|
+          opts.banner = BANNER
+          opts.on('--certname x,y,z', Array, 'the name(s) of the cert(s) to be signed') do |cert|
+            parsed['certname'] = cert
+          end
+          opts.on('--help', 'Display this sign specific help output') do |help|
+            parsed['help'] = true
+          end
+          opts.on('--version', 'Output the version') do |v|
+            parsed['version'] = true
+          end
+          opts.on('--all', 'Operate on all certnames') do |a|
+            parsed['all'] = true
+          end
+        end
+      end
+
+      def initialize(logger)
+        @logger = logger
+      end
+
+      def run(input)
+        if input['all']
+          requested_certnames = get_all_pending_certs
+          if requested_certnames.nil?
+            return 1
+          end
+        else
+          requested_certnames = input['certname']
+        end
+
+        exit_code = sign_requested_certs(requested_certnames)
+        return exit_code
+      end
+
+      def get_certificate_statuses
+        uri = URI('https://localhost:8140/puppet-ca/v1/certificate_statuses/:any_key?environment=production')
+        Net::HTTP.start(uri.host, uri.port, use_ssl: true, cert_store: store) do |connection|
+          request = Net::HTTP::Get.new uri
+          connection.request(request)
+        end
+      end
+
+      def sign_cert(certname)
+        uri = URI('https://localhost:8140/puppet-ca/v1/certificate_status/' + certname + '?environment=production')
+        Net::HTTP.start(uri.host, uri.port, use_ssl: true, cert_store: store ) do |connection|
+          request = Net::HTTP::Put.new(uri, HEADERS)
+          request.body = BODY
+          connection.request(request)
+        end
+      end
+
+      def get_all_certs
+        result = get_certificate_statuses
+        unless result.is_a? Net::HTTPOK
+            @logger.err 'Error:'
+            @logger.err "    #{result.inspect}"
+            return nil
+        end
+        return result
+      end
+
+      def select_pending_certs(get_result)
+        requested_certnames = JSON.parse(get_result.body).select{|e| e["state"] == "requested"}.map{|e| e["name"]}
+
+        if requested_certnames.empty?
+          @logger.err 'Error'
+          @logger.err "No waiting certificate requests to sign"
+          return nil
+        end
+
+        return requested_certnames
+      end
+
+      def get_all_pending_certs
+        result = get_all_certs
+        if result
+          select_pending_certs(result)
+        end
+      end
+
+      def sign_requested_certs(certnames)
+        exit_code = 0
+        certnames.each do |certname|
+          result = sign_cert(certname)
+          case result.code
+          when '204'
+            @logger.inform "Signed certificate for #{certname}"
+          when '404'
+            @logger.err 'Error'
+            @logger.err "    Could not find certificate for #{certname}"
```

Comment:
The error output here varies between in whitespace and colon usage. I think we should circle back around after this and have them all conform to a single format, but for the time being we should make these all the same per action.

I think in this PR we have `Error` with and without a following colon, and then the following line with or without four spaces of indentation.